internal/addoninstancesdk/sdk.go

```diff
@@ -0,0 +1,152 @@
+	isRunning      bool
+	isRunningMutex *sync.Mutex
+
+	// the latest condition which the heartbeat reporter would be reporting the periodically
+	latestCondition metav1.Condition
+
+	// to control the rate at which the heartbeat reporter would run
+	currentInterval time.Duration
+	ticker          *time.Ticker
+
+	// for effectively communicating the stop and update signals
+	stopperCh chan bool
+	updateCh  chan updateOptions
+}
+
+// ensure that the `AddonInstanceHeartbeatReporter` implements the `AddonInstanceStatusReporterClient` interface
+var _ AddonInstanceStatusReporterClient = (*AddonInstanceHeartbeatReporter)(nil)
+
+// InitializeAddonInstanceHeartbeatReporterSingleton sets up a singleton of the type `AddonInstanceHeartbeatReporter` (only if it doesn't exist yet) and returns it to the caller.
+func InitializeAddonInstanceHeartbeatReporterSingleton(addonInstanceInteractor AddonInstanceInteractorClient, addonName string, addonTargetNamespace string) (*AddonInstanceHeartbeatReporter, error) {
+	if addonInstanceHeartbeatReporterSingleton == nil {
+		addonInstanceHeartbeatReporterSingletonMutex.Lock()
+		defer addonInstanceHeartbeatReporterSingletonMutex.Unlock()
+		if addonInstanceHeartbeatReporterSingleton == nil {
+			addonInstanceHeartbeatReporterSingleton = &AddonInstanceHeartbeatReporter{
+				AddonInstanceInteractor: addonInstanceInteractor,
+				AddonName:               addonName,
+				AddonTargetNamespace:    addonTargetNamespace,
+				isRunning:               false,
+				isRunningMutex:          &sync.Mutex{},
+				latestCondition: metav1.Condition{
+					Type:    "addons.managed.openshift.io/Healthy",
+					Status:  "False",
+					Reason:  "NoHeartbeatReported",
+					Message: fmt.Sprintf("Addon '%s' hasn't reported any heartbeat yet", addonName),
+				},
+				stopperCh: make(chan bool),
+				updateCh:  make(chan updateOptions),
+			}
+
+			currentAddonInstance, err := addonInstanceHeartbeatReporterSingleton.AddonInstanceInteractor.GetAddonInstance(context.TODO(), "addon-instance", addonInstanceHeartbeatReporterSingleton.AddonTargetNamespace)
+			if err != nil {
+				return nil, fmt.Errorf("error occurred while fetching the current heartbeat update period interval")
+			}
+			addonInstanceHeartbeatReporterSingleton.currentInterval = currentAddonInstance.Spec.HeartbeatUpdatePeriod.Duration
+		}
+	}
+
+	return addonInstanceHeartbeatReporterSingleton, nil
+}
+
+func GetAddonInstanceHeartbeatReporterSingleton() (*AddonInstanceHeartbeatReporter, error) {
+	if addonInstanceHeartbeatReporterSingleton == nil {
+		return nil, fmt.Errorf("heartbeat-reporter not found to be initialised. Initialize it by calling `InitializeAddonInstanceHeartbeatReporterSingleton(...)`")
+	}
+	return addonInstanceHeartbeatReporterSingleton, nil
+}
+
+func (adihrClient *AddonInstanceHeartbeatReporter) changeRunningState(desiredState bool) {
+	adihrClient.isRunningMutex.Lock()
+	adihrClient.isRunning = desiredState
+	adihrClient.isRunningMutex.Unlock()
```

Comment:
> Warning: The suggestion is not ordered correctly.

Why so? In your suggestion (and in the original implementation), the lock seems to wrap the critical section fairly. Anything I am missing here?

> This function is also not closing/emitting on the stopperCh, leaving clients stuck.

Do you mean to say that this function should be `stopperCh` aware and tweak the stopperCh accordingly? If yes, then I think that would unnecessarily complicated because the updates to `isRunning` attribute are strongly tied to the `StatusReporter.Start(...)` method wherein the `isRunning` attribute is already being handled safely (turning it to `true` in the beginning and deferring it to `false` whenever the method exits).